openupgrade_scripts/scripts/stock/16.0.1.1/pre-migration.py

```diff
@@ -0,0 +1,131 @@
+    ("stock.move.line", "stock_move_line", "product_uom_qty", "reserved_uom_qty"),
+    ("stock.rule", "stock_rule", "location_id", "location_dest_id"),
+    ("stock.picking", "stock_picking", "move_lines", "move_ids"),
+]
+
+
+def _update_stock_quant_storage_category_id(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE stock_quant
+        ADD COLUMN IF NOT EXISTS storage_category_id INTEGER
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE stock_quant
+        SET storage_category_id = stock_location.storage_category_id
+        FROM stock_location
+        WHERE stock_quant.location_id = stock_location.id
+        """,
+    )
+
+
+def _update_sol_product_category_name(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE stock_move_line
+        ADD COLUMN IF NOT EXISTS product_category_name CHARACTER VARYING
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE stock_move_line AS sml
+        SET product_category_name = category.complete_name
+        FROM product_product AS product
+        JOIN product_template AS product_tmpl
+        ON product_tmpl.id = product.product_tmpl_id
+        JOIN product_category AS category
+        ON category.id = product_tmpl.categ_id
+        WHERE sml.product_id = product.id
+        """,
+    )
+
+
+def _compute_stock_location_replenish_location(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE stock_location
+        ADD COLUMN IF NOT EXISTS replenish_location BOOLEAN
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        WITH location_info as (
+            SELECT sl.id as id,
+            CASE
+                WHEN sl.usage = 'internal' AND sl.id = sw.lot_stock_id THEN True
+                ELSE FALSE
+            END as replenish_location_status
+            FROM stock_location sl
+            LEFT JOIN stock_warehouse sw
+            ON sw.id = sl.warehouse_id
+        )
+        UPDATE stock_location sl
+        SET replenish_location = info.replenish_location_status
+        FROM location_info info
+        WHERE sl.id = info.id
+        """,
+    )
+
+
+def _update_stock_quant_package_pack_date(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE stock_quant_package
+        ADD COLUMN IF NOT EXISTS pack_date DATE
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE stock_quant_package
+        SET pack_date = DATE(create_date)
+        """,
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.rename_tables(env.cr, _tables_renames)
+    openupgrade.rename_models(env.cr, _models_renames)
+    openupgrade.rename_fields(env, _fields_renames)
+    openupgrade.rename_columns(env.cr, _columns_renames)
```

Comment:
```suggestion
    openupgrade.rename_fields(env, _fields_renames)
```
Dùng mỗi `rename_fields` là được mà anh, thêm `rename_column` sẽ bị thừa và nó cũng gây lỗi trên con bot kìa a

Reply:
a đọc qua thì k thấy rename_fields nó rename column trong bảng

Reply:
nó đang giống rename_model và rename_table vậy

Reply:
Có mà a zai, chi tiết ở https://github.com/OCA/openupgradelib/blob/66c1f84dc0c4f3d187283efec9af98d8e3ddc4a4/openupgradelib/openupgrade.py#L660

Reply:
oh, nó ở ngay đầu. ok bro :smile:

Reply:
Yes sir, với cả a nên chạy thử migrate và quay video cho nhưng con core to to như này, DB đang dùng để chạy là DB chỉ cắm addon Odoo và Tvt ngoài ko có bất kỳ addons nào khác, và cũng chỉ cài full Module của tvt thôi

Reply:
À lộn DB bên tvtaddon upgrade, chỗ này chỉ addon odoo thôi :v

Reply:
ok bro, a đang up code lên trc r dựng thử